src/wallet/coinselection.cpp

```diff
@@ -0,0 +1,309 @@
+            break;
+        }
+        // Conditions for starting a backtrack
+        else if ( value_track + lookahead < target_value || // Cannot possibly reach target with the amount remaining in the lookahead.
+                    value_track > target_value + cost_of_change || // Selected value is out of range, go back and try other branch
+                    (curr_waste > best_waste && (utxo_pool.at(0).fee - utxo_pool.at(0).long_term_fee) > 0)) { // Don't select things which we know will be more wasteful if the waste is increasing
+            backtrack = true;
+        } else if (value_track >= target_value) { // Selected value is within range
+            curr_waste += (value_track - target_value); // This is the excess value which is added to the waste for the below comparison
+            // Adding another UTXO after this check could bring the waste down if the long term fee is higher than the current fee.
+            // However we are not going to explore that because this optimization for the waste is only done when we have hit our target
+            // value. Adding any more UTXOs will be just burning the UTXO; it will go entirely to fees. Thus we aren't going to
+            // explore any more UTXOs to avoid burning money like that.
+            if (curr_waste <= best_waste) {
+                best_selection.assign(selection.begin(), selection.end());
+                best_waste = curr_waste;
+            }
+            curr_waste -= (value_track - target_value); // Remove the excess value as we will be selecting different coins now
+            backtrack = true;
+        }
+
+        // Backtracking, moving backwards
+        if (backtrack) {
+            backtrack = false; // Reset
+            --depth;
+
+            // Walk backwards to find the first utxo which has not has its second branch traversed
+            while (!selection.at(depth)) {
+                selection.at(depth) = false;
+                selection.at(depth) = false;
+                lookahead += utxo_pool.at(depth).txout.nValue;
+
+                // Step back one
+                --depth;
+
+                if (depth < 0) { // We have walked back to the first utxo and no branch is untraversed. No solution, exit.
+                    done = true;
+                    break;
+                }
+            }
+
+            if (!done) {
+                // These were always included first, try excluding now
+                selection.at(depth) = false;
+                value_track -= utxo_pool.at(depth).txout.nValue;
+                curr_waste -= (utxo_pool.at(depth).fee - utxo_pool.at(depth).long_term_fee);
+                ++depth;
+            }
+        } else { // Moving forwards, continuing down this branch
+            // Assert that this utxo is not negative. It should never be negative, effective value calculation should have removed it
+            assert(utxo_pool.at(depth).txout.nValue >= 0);
```

Comment:
I thought you're filtering effective values to be strictly bigger than zero.

Reply:
Fixed